### PR TITLE
[codenav] adding prototypes to  codenav

### PR DESCRIPTION
**Pull request by Numbers88s**
This is a draft PR to address the problem with our current implementation of "Find Implementations" as described in [this issue](https://github.com/sourcegraph/sourcegraph/issues/50110).

>Currently, "Find implementations" shows the prototype instead of implementations. When a symbol A# implements symbol B# then I expect "Find implementations" on the symbol B# to return A#.

Feedback welcome, but more importantly need to ensure this actually solves the problem highlighted by TwoSigma.

animalsExample

## Test plan
Manually tested the affected routes and responses to ensure no regressions. Updated tests as well.

Please review the changes and provide any feedback or suggestions. If there are any concerns or further adjustments needed, let me know and I'll address them promptly.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.02% (+2.23 kb) | 0.02% (+2.23 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/cesar/codenav/real-add-prototypes/compare-report.html;tab=live_object) for a full comparison between the commits 2b4a5b6 and 1be59cf or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by efritz** [comment]
Fixes #51148.

File: enterprise/internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go
Comment (efritz) [inline enterprise/internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go:1]: Do we have a new test case for the prototypes field?

File: enterprise/internal/codeintel/codenav/service_implementations_test.go [hunk 1]
```diff
@@ -284,12 +284,9 @@ func TestImplementationsRemote(t *testing.T) {
 		}
 	}
 
-	if history := mockLsifStore.GetBulkMonikerLocationsFunc.History(); len(history) != 4 {
```
Comment (efritz) [inline enterprise/internal/codeintel/codenav/service_implementations_test.go:287]: Same here, not sure if we're only testing impls and not prototypes?

**Review by efritz — COMMENTED** [review]
LGTM so far without testing locally, let's see what @olafurpg says about the new experience, then I'm good to :shipit:

**Comment by olafurpg** [comment]
I'm sorry for the slow review. I'm super excited to see this issue get fixed!

The "Prototypes" section currently doesn't behave like I expect it to. For example, the prototype list of the toplevel symbol (the abstract field on the interface) should be empty.

CleanShot 2023-04-26 at 16 41 52@2x


I also noticed that the "Implementations" section shows a result that is not a definition

CleanShot 2023-04-26 at 16 44 01@2x


Here's the SCIP file that I uploaded for the scip-java codebase [index.scip.zip](https://github.com/sourcegraph/sourcegraph/files/11334186/index.scip.zip) (note that the paths point to my local file system, but you can snapshot it locally with `scip snapshot --project-root=$PWD`).

**Comment by Numbers88s** [comment]
Recent changes (cc @olafurpg):

The prototype list of the toplevel symbol (the abstract field on the interface) is now correctly shown as empty.

ProtoEmpty

In addition, the "Implementations" section does not show references.

UpdatedReference

And lastly, I've updated the initial image displaying the original mock repo with its correct separation of implementation and prototypes.

animalsExample

File: enterprise/internal/codeintel/codenav/service.go [hunk 2]
```diff
@@ -719,6 +719,123 @@ func (s *Service) GetImplementations(ctx context.Context, args RequestArgs, requ
...
+	implementationLocations, err := s.getUploadLocations(ctx, args, requestState, locations, true)
+	if err != nil {
+		return nil, cursor, err
+	}
+	trace.AddEvent("TODO Domain Owner", attribute.Int("numImplementationsLocations", len(implementationLocations)))
+
+	return implementationLocations, cursor, nil
+}
+
+func (s *Service) GetPrototypes(ctx context.Context, args RequestArgs, requestState RequestState, cursor ImplementationsCursor) (_ []shared.UploadLocation, _ ImplementationsCursor, err error) {
```
[the remark this answers is not in the document]
Comment (Numbers88s) [reply enterprise/internal/codeintel/codenav/service.go:774]: Yeah its a weird diff, but I've removed the second phase altogether for prototypes.

File: enterprise/internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go
Comment (Numbers88s) [reply enterprise/internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go:1]: I've added new test cases for prototypes.

**Review by efritz — APPROVED** [review]

**Comment by olafurpg** [comment]
I tried out the latest commit and it's a huge improvement! I noticed that the symbol is included as an implementation of itself if there are no true implementations. I'd expect it to be empty instead

CleanShot 2023-05-03 at 14 37 39@2x

It works correctly when there are true implementations, then it doesn't show the symbol itself as an implementation

CleanShot 2023-05-03 at 14 40 31@2x

**Comment by Numbers88s** [comment]
It seems that by removing Phase 2, we're able to fix the problem with the misplaced implementation. Good catch!

CleanShot 2023-05-08 at 12 46 08

I'll wrap things up for now. If any other issues pop up, we can always tackle them in a follow-up pull request. Just let me know!

**Pull request merged**